### DISCUSSION TITLE
Error 2 following make command for hlrdas/noah-mp module

### DISCUSSION THREAD

**Discussion by psbecker** [NCAR/hrldas#194] [Q&A] [answered]
Hi all,
I am following the hlrdas/ noah-mp tutorial. and after updating the paths and typing the "make" command, I get the following error but am unsure what it means. But the executables are not there 

`(cd Utility_routines;		make)
make[1]: Nothing to be done for `all'.
(cd ../noahmp/utility;          make)
make[1]: Nothing to be done for `all'.
(cd ../noahmp/src;              make)
make[1]: *** No rule to make target `../drivers/hrldas/NoahmpIOVarType.o', needed by `GroundWaterMmfMod.o'.  Stop.
make: *** [all] Error 2
`

**Comment by cenlinhe** [comment]
What computer machine are you running the model? What compiler did you use? Looks like this is related to the computer environment you set up.

**Comment by psbecker** [comment]
I am running it on a Mac and used the Linux gfortran compiler serial option. I updated the pths in the 'user_build_options'
________________________________
From: Cenlin_He ***@***.***>
Sent: Monday, July 8, 2024 10:02 PM
To: NCAR/hrldas ***@***.***>
Cc: Paige Becker ***@***.***>; Author ***@***.***>
Subject: [EXTERNAL] Re: [NCAR/hrldas] Error 2 following make command for hlrdas/noah-mp module (Discussion #194)

CAUTION: This email originated from outside of the Colorado School of Mines organization. Do not click on links or open attachments unless you recognize the sender and know the content is safe.


What computer machine are you running the model? What compiler did you use? Looks like this is related to the computer environment you set up.

**Reply by psbecker** [reply]
Tried re-compiling it and got the following Error: 
```
(cd Utility_routines;		make)
make[1]: Nothing to be done for `all'.
(cd ../noahmp/utility;          make)
make[1]: *** No rule to make target `Machine.o', needed by `all'.  Stop.
make: *** [all] Error 2
```
My user_build_options has the following:
`#=============================================================================================
#  Options for Linux with gnu gfortran serial
#  Please change the library paths according to your computer setup
#=============================================================================================

 COMPILERF90    =       gfortran
 FREESOURCE     =       -ffree-form  -ffree-line-length-none
 F90FLAGS       =       -g -fconvert=big-endian -fbounds-check -fno-range-check -fallow-argument-mismatch
 MODFLAG        =       -I
 LDFLAGS        =
 CPP            =       cpp
 CPPFLAGS       =       -P -traditional -D_GFORTRAN_ # -DSPATIAL_SOIL
 LIBS           =
 LIBJASPER      =      -ljasper
 INCJASPER      =      -I/opt/homebrew/Cellar/jasper/4.2.4/include
 NETCDFMOD      =      -I/opt/homebrew/Cellar/netcdf-fortran/4.6.1/include
 NETCDFLIB      =      -L/opt/homebrew/Cellar/netcdf-fortran/4.6.1/lib -lnetcdf -lnetcdff
 BZIP2          =       NO
# BZIP2_INCLUDE  =       -I/usr/include
# BZIP2_LIB      =       -L/usr/lib64 -lbz2
 RM             =       rm -f
 CC             =       cc
`

**Reply by cenlinhe** [reply]
We have not run the model directly on a MAC system before, but we have this Docker container that already included all the required packages to run the model: https://github.com/NCAR/hrldas/blob/develop/tutorial/NoahMP_docker_container_setup.pdf
You could download the docker container and run in on your Mac.
Here is a tutorial for running the model in the docker container: https://github.com/NCAR/hrldas/blob/develop/tutorial/Note6_AMS2024_NoahMP_short_course.ipynb

Anyone has any suggestions on this issue for running the model directly on a Mac system?

**Reply by psbecker** [reply]
I tried the docker container version and get a similar error even though I've checked the file paths multiple times.

**Reply by cenlinhe** [reply, marked as answer]
this is weird. We have tested the container in the Mac system which works. Did you choose "9. Linux gfortran/gcc compiler MPI (Docker container)" when you configure the model (./configure)? After selecting this option, you do not need to change any software path inside the user_build_option file since it is already set to work with the container environment.

**Reply by psbecker** [reply]
Oh okay that makes more sense. I think I misread the tutorial and thought I needed to change the paths. But it was able to run the single point file executable!

**Reply by gigachad1306** [reply]
@cenlinhe  @psbecker, may you please help me with the installation of Jasper? I am encountering some errors.

**Comment by gigachad1306** [comment]
I am facing an issue with the installation of the Jasper Library.
Kindly help me with that.
I have successfully downloaded and installed the Netcdf library.

@cenlinhe

**Reply by cenlinhe** [reply]
Hi there, the installation of Jasper library varies across computer systems and platforms, which needs you to talk to your technical staff to help you, because we do not know your computer system and platform. You may need to google the specific steps to install Jasper in your specific computer system. If you cannot resolve the Jasper library issue on your computer, you can directly download and install our Noah-MP docker container which already includes all the required library to run the model: https://github.com/NCAR/hrldas/blob/master/tutorial/NoahMP_docker_container_setup.pdf

**Comment by gigachad1306** [comment]
make[1]: Leaving directory '/data/Devesh/hrldas/urban/wrf'
(cd IO_code;			make)
make[1]: Entering directory '/data/Devesh/hrldas/hrldas/IO_code'

rm -f module_hrldas_netcdf_io.f90
cpp -P -traditional -D_GFORTRAN_  -D_HRLDAS_OFFLINE_  -I/usr/include module_hrldas_netcdf_io.F > module_hrldas_netcdf_io.f90
module_hrldas_netcdf_io.F:2599:2: fatal error: netcdf.inc: No such file or directory
 2599 | 
      |  ^           
compilation terminated.
make[1]: *** [Makefile:39: module_hrldas_netcdf_io.o] Error 1
make[1]: Leaving directory '/data/Devesh/hrldas/hrldas/IO_code'
make: *** [Makefile:13: all] Error 2



**after installing netcdf and jasper. Now I got this error.
kindly help**

**Reply by cenlinhe** [reply]
This error is because you did not set the netcdf path correctly in your user_build_option file before compiling.

**Comment by gigachad1306** [comment]
**In a different comuter system the following error is coming:**

make[1]: Leaving directory '/mnt/d/Devesh_PhD/hrldas/hrldas/Utility_routines'
(cd ../noahmp/utility;          make)
make[1]: Entering directory '/mnt/d/Devesh_PhD/hrldas/noahmp/utility'
make[1]: *** No rule to make target 'Machine.o', needed by 'all'.  Stop.
make[1]: Leaving directory '/mnt/d/Devesh_PhD/hrldas/noahmp/utility'
make: *** [Makefile:9: all] Error 2

**Reply by cenlinhe** [reply]
This error seems similar to the one above, which is because you did not set the netcdf and compiler path correctly in your user_build_option file before compiling.

**Comment by gigachad1306** [comment]
**path mentioned in the user_build_option file:**

 NETCDFMOD  =      -I/usr/include
 NETCDFLIB      =      -L/usr/lib/x86_64-linux-gnu -lnetcdf -lnetcdff 

----------------------------------------------------------------------------------

**I used the path mentioned here below (in bold):
nc-config --all**
------------------------------------------------------------------------------------------
This netCDF 4.9.2 has been built with the following features: 

  --cc            -> /usr/bin/cc
  --cflags        -> -I/usr/include -I/usr/include/hdf5/serial
  --libs          -> -L/usr/lib/x86_64-linux-gnu -L/usr/lib/x86_64-linux-gnu/hdf5/serial -lnetcdf
  --static        -> -lhdf5_hl -lhdf5 -lcrypto -lcurl -lpthread -lsz -lz -ldl -lm -lxml2


  --has-dap          -> yes
  --has-dap2         -> yes
  --has-dap4         -> yes
  --has-nc2          -> yes
  --has-nc4          -> yes
  --has-hdf5         -> yes
  --has-hdf4         -> no
  --has-logging      -> no
  --has-pnetcdf      -> no
  --has-szlib        -> no
  --has-cdf5         -> yes
  --has-parallel4    -> no
  --has-parallel     -> no
  --has-nczarr       -> yes
  --has-zstd         -> yes
  --has-benchmarks   -> no
  --has-multifilters -> no
  --has-stdfilters   -> deflate szip bz2
  --has-quantize     -> no
  --prefix        -> /usr

  **--includedir    -> /usr/include
  --libdir        -> /usr/lib/x86_64-linux-gnu**

  --plugindir     -> 
  --version       -> netCDF 4.9.2
-------------------------------------------------------------------------------------------------------

Could you please confirm if it is correct and how I can rectify the mistake and resolve the error?

**Reply by cenlinhe** [reply]
your installation for the path looks correct. You can double check by going to /usr/include and /usr/lib/x86_64-linux-gnu to see if the installed netcdf files are there. Particularly, based on your earlier error message, it looks like the file "netcdf.inc" is missing.

**Reply by gigachad1306** [reply]
Yes, the error, file "netcdf.inc" is missing is still there. How to resolve it?

**Reply by cenlinhe** [reply]
@tslin2 @CharlesZheZhang Do you have any suggestions? My feeling is that this issue is still somehow related to the installation and environment setup for the software packages.

**Reply by tslin2** [reply]
Is your netcdf built with Fortran library?
please try to build NetCDF with Fortran such as https://docs.unidata.ucar.edu/netcdf-c/current/building_netcdf_fortran.html

**Reply by tslin2** [reply]
Please check whether NetCDF Fortran components have been installed:
nf-config --all

**Reply by gigachad1306** [reply]
the netcdf.inc issue is resolved. Now the following error is coming up:


make[3]: Entering directory '/data/Devesh/hrldas/hrldas/HRLDAS_forcing/lib'
make[3]: Nothing to be done for 'all'.
make[3]: Leaving directory '/data/Devesh/hrldas/hrldas/HRLDAS_forcing/lib'
gfortran -o create_forcing.exe -I./lib -g -fconvert=big-endian -fbounds-check -fno-range-check -fallow-argument-mismatch -I./lib create_forcing.o \
	-L./lib -lsmda -L/usr/local/lib -lnetcdf -lnetcdff   -L/usr/local/lib -ljasper
/usr/bin/ld: ./lib/libsmda.a(decode_jpeg2000.o): in function `info_jpeg2000_':
decode_jpeg2000.c:(.text+0xd9): undefined reference to `jpc_decode'
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:29: create_forcing.exe] Error 1
make[2]: Leaving directory '/data/Devesh/hrldas/hrldas/HRLDAS_forcing'
make[1]: *** [Makefile:13: default] Error 2
make[1]: Leaving directory '/data/Devesh/hrldas/hrldas/HRLDAS_forcing'
make: *** [Makefile:15: all] Error 2

**Reply by tslin2** [reply]
This looks like jasper issue,
can you try this one
 https://www2.mmm.ucar.edu/wrf/OnLineTutorial/compile_tutorial/tar_files/jasper-1.900.1.tar.gz

you can find other version here: https://github.com/jasper-software/jasper/releases

**Reply by tslin2** [reply]
you may also refer to this issue (https://github.com/wrf-model/WPS/issues/207)

**Reply by gigachad1306** [reply]
> This looks like jasper issue, can you try this one https://www2.mmm.ucar.edu/wrf/OnLineTutorial/compile_tutorial/tar_files/jasper-1.900.1.tar.gz
> 
> you can find other version here: https://github.com/jasper-software/jasper/releases

I tried with 4.2.5 and 1.900.1

This is the error:




make[2]: Leaving directory '/data/Devesh/hrldas/hrldas/HRLDAS_forcing/lib'
(make -f Makefile all)
make[2]: Entering directory '/data/Devesh/hrldas/hrldas/HRLDAS_forcing'

rm -f create_forcing.f90
cpp -P -traditional -D_GFORTRAN_   create_forcing.F > create_forcing.f90
gfortran -ffree-form  -ffree-line-length-none -g -fconvert=big-endian -fbounds-check -fno-range-check -fallow-argument-mismatch -c -I/usr/local/include -I./lib -I./lib create_forcing.f90
(cd lib; make)
make[3]: Entering directory '/data/Devesh/hrldas/hrldas/HRLDAS_forcing/lib'
make[3]: Nothing to be done for 'all'.
make[3]: Leaving directory '/data/Devesh/hrldas/hrldas/HRLDAS_forcing/lib'
gfortran -o create_forcing.exe -I./lib -g -fconvert=big-endian -fbounds-check -fno-range-check -fallow-argument-mismatch -I./lib create_forcing.o \
	-L./lib -lsmda -L/usr/local/lib -lnetcdf -lnetcdff   -L/usr/local/lib -ljasper
/usr/bin/ld: ./lib/libsmda.a(decode_jpeg2000.o): in function `info_jpeg2000_':
decode_jpeg2000.c:(.text+0xd5): undefined reference to `jpc_decode'
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:29: create_forcing.exe] Error 1
make[2]: Leaving directory '/data/Devesh/hrldas/hrldas/HRLDAS_forcing'
make[1]: *** [Makefile:13: default] Error 2
make[1]: Leaving directory '/data/Devesh/hrldas/hrldas/HRLDAS_forcing'
make: *** [Makefile:15: all] Error 2
